server/update-workflow-interface/update-workflow-interface.md

```diff
@@ -0,0 +1,217 @@
+
+Similar to the uncertain duration, it may be know a priori that the update will
+take more time than the issuer is willing to block the thread of execution
+invoking the update. In such cases the update issuer would prefer to not wait at
+all for the UpdateResponse, preferring instead to poll for it later.
+
+### Invalid Update
+
+UpdateRequests can be invalid given the state of the receiving workflow. In such
+cases the UpdateResponse will contain an error indication and payload. Invalid
+updates _must not_ be recorded in workflow history to prevent a malicious issuer
+from causing a denial of service on the workflow by issuing repeated invalid
+UpdateRequests.
+
+### "Failing" Update
+
+UpdateRequests can be semantically acceptable but still fail due to external
+dependencies. In such cases the UpdateResponse will contain an error indication
+and payload.
+
+### Unexpected Disconnect
+
+While the issuer of an update may wish to block on an UpdateResponse, network
+conditions or software failures may occur which forcibly disconnects the update
+issuer from the server at the time that the Update completes _and_ before the
+issuer becomes aware of the Update ID. The issuer must be able to re-issue the
+same UpdateRequest to access the server-generated Update ID.
+
+## gRPC API
+
+The gRPC API comprises two functions, `UpdateWorkflow` and
+`AwaitUpdateWorkflowResponse` to be added to the `WorkflowService` service
+definition. The former is used to invoke an update and follows the standard
+patterns for identifying a workflow execution and carrying arbitrary metadata
+(Header) and payloads (Payloads). The request type for this function includes a
+`wait_timeout` property indicating how long the client is willing to wait for a
+synchronous response once the request has been accepted by the system. This
+value can be set to zero indicating no such waiting should occur.  The
+`request_id` property uniquely identifies the logical request (_not_ the
+issuance of the request) and is intended for use as an idempotency token.
+
+The second function, `AwaitUpdateWorkflowResponse` is used when a client has
+already issued an `UpdateWorkflow` request and has the resulting `update_id`.
+That `update_id` can be provided along with a `wait_timeout` (same semantics as
+above) to re-attach to an in-flight or completed Update and obtain the
+associated UpdateResult in the form of an `UpdateWorkflowResponse`.
+
+
+```protobuf
+service WorkflowService {
+    rpc UpdateWorkflow(UpdateWorkflowRequest) returns (UpdateWorkflowResponse){}
+    rpc AwaitUpdateWorkflowResponse(AwaitUpdateWorkflowResponseRequest) returns (UpdateWorkflowResponse){}
+}
+
+message UpdateWorkflowRequest {
+    temporal.api.common.v1.Header header = 1;
+    string request_id = 2;
+    string namespace = 3;
+    temporal.api.common.v1.WorkflowExecution execution = 4;
+    string update_name = 5;
+    temporal.api.common.v1.Payloads input = 6;
+    google.protobuf.Duration wait_timeout = 7;
```

Comment:
The reason to have this apart from the gRPC timeout is so that the server can return an empty-ish response after `wait_timeout` that includes the `update_id`. From the gRPC perspective that's a successful call/response pair but at a higher layer we understand that there is more to do to see the UpdateResponse. Giving back the `update_id` is necessary so that the client can come look for it later.

Reply:
I see.

I wonder then if it's clearer from a code POV to always require wait for response as a separate call and have this call always be fast? Basically we don't allow `StartWorkflowExecution` to wait on its response. So have this function always return the update ID immediately and always make obtaining the result use another call. Internally on the server I figure they share that kind of "awaiting" code anyways, and no need for shortcuts for the SDK's benefit (it's actually easier from an SDK to have only one way to wait on a result).

Reply:
> Can't always be fast because we want it to return the UpdateID

It could if the worker wasn't in the way :-) But I understand. I don't have a strong opinion here. To me a send request and get response is a very natural separation for Temporal (especially if sending a request doesn't block until the worker accepts it).

Reply:
I'm foreseeing issues where the request deadline is shorter than wait_timeout, can we just use the deadline for this purpose?

Reply:
I think service can take min of the deadline and wait_timeout. Is the deadline causes a client to cut the request or it is purely implemented by the server side?

Reply:
Client must throw if deadline exceeds, it might never get a response from the server..

Reply:
I wanted to use the `wait_timeout` field to reinforce conceptually that there is a difference between "how long this RPC might run" and "how long I want to wait for an UpdateResponse." However it doesn't seem to be clarifying very much and may be having the opposite effect. I can proceed using just the built-in gRPC timeout and have that _implicitly_ be the UpdateResponse wait interval.

Reply:
> Having one RPC call to send request and get response is critical for low-latency use case

I think it may be good to clarify the server logic here. Is the update-and-wait code not internally equivalent to calling request update then calling wait? If so, the only latency concern is a separate gRPC invocation? If I want really low latency, can I disable update validation?

Reply:
> Is the update-and-wait code not internally equivalent to calling request update then calling wait?

The intent is for those two approaches to be functionally equivalent. If I've made it seem otherwise then I messed something up.

> If I want really low latency, can I disable update validation?

This is rather a longer conversation but it's the closes thing to having our cake and eating it too. If we were somehow able to inform the history service that a given Update path has no validation for all running versions then we could optimize away the validate/execute two-step. As it stands AFAIK the sever has no understanding of the structure of Queries/Signals supported by a workflow (maybe I'm wrong on this) but in theory we could generate an IDL-like structure at runtime so that the server could be aware of the existence of validation (along with other things like the name bindings of Update functions)

Reply:
To clarify, I don't mind not having such a skip-validation option, but I just thought it was strange we're afraid of two gRPC calls due to low-latency, but we're ok with two worker calls.

Reply:
After further discussion, I have been told two important things:

1. We will not require a worker be running a workflow to accept an update from a client
2. We will not separate the update validation and update execution steps across a server call boundary, only worker-local separation if any

Given these, all my concerns go away (can resolve this conversation, leaving open for my response to be visible).

Reply:
I no longer have a strong opinion on whether this individual call is a long polling call. Resolving.